apollo-router/src/plugins/request_decompression.rs

```diff
@@ -0,0 +1,90 @@
+use crate::plugin::{PluginInit, PluginPrivate};
+use crate::services::router;
+use schemars::JsonSchema;
+use serde::{Deserialize, Serialize};
+use tower::{ServiceBuilder, ServiceExt};
+
+#[derive(Debug, Default, Deserialize, JsonSchema, Serialize)]
+struct RequestDecompressionConfig {
+    /// When `true`, brotli-encoded request bodies will be automatically decompressed.
+    ///
+    /// Defaults to `true`.
+    br: Option<bool>,
+    /// When `true`, gzip-encoded request bodies will be automatically decompressed.
+    ///
+    /// Defaults to `true`.
+    gzip: Option<bool>,
+    /// When `true`, deflate-encoded request bodies will be automatically decompressed.
+    ///
+    /// Defaults to `true`.
+    deflate: Option<bool>,
+}
+struct RequestDecompressionPlugin {
+    br: bool,
+    gzip: bool,
+    deflate: bool,
+}
+
+#[async_trait::async_trait]
+impl PluginPrivate for RequestDecompressionPlugin {
+    type Config = RequestDecompressionConfig;
+
+    async fn new(init: PluginInit<Self::Config>) -> Result<Self, tower::BoxError> {
+        Ok(RequestDecompressionPlugin {
+            br: init.config.br.unwrap_or_default(),
+            gzip: init.config.gzip.unwrap_or_default(),
+            deflate: init.config.deflate.unwrap_or_default(),
+        })
+    }
+
+    fn router_service(&self, service: router::BoxService) -> router::BoxService {
+        // RequestDecompressionLayer works on `http` types so we have to perform the conversions to
+        // make it compatible with `BoxService`.
+        ServiceBuilder::new()
+            .map_request(http::Request::from)
+            .map_response(router::Response::from)
+            .layer(
```

Comment:
We're probably going to need an `map_err` here. All responses should look graphql shaped.

Reply:
Let's talk soon about how to do this the best way. I'm not familiar enough with this yet to do it on my own.